2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,348 @@
+currently published to PowerShellGallery.com) is ~700KB in compressed json form.
+
+The cache would have both latest stable and latest prerelease versions of resources.
+
+>[!NOTE]
+>Need to experiment if it makes sense to have a single cache file for a repositories
+>or a different cache per repository for size and perf reasons.
+>Perf tests will determine if the cache needs to be in binary form.
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+This means that there is no local cache of that repository and operations will
+always connect to that repository.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v3` as
+that will be assumed and discovered at runtime (trying v3 first, then falling
+back to v2, then the literal URL).
+Support for local filesystem repositories must be maintained.
+A `-Trusted` switch indicates whether to prompt the user when installing resources
+from that repository.
+By default, if this switch is not specified, the repository is untrusted.
+A `-Repositories` parameter will accept an array of hashtables equivalent to
+the parameter names.
+A `-Name` parameter allows for setting a friendly name.
+
+A `-Default` switch will set one repository as the default (when not specified
+with other cmdlets).
+Each time it is specified, it sets the new one as default and the previous default
+is no longer default.
+
+A `-Scope` parameter will support `AllUsers` and `CurrentUsers`.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL, trust level,
+or if that repository is the default.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+For each repository, if the local cache does not exist, the cmdlet will call
+`Update-PSResourceCache` first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
```

Comment:
`Update-PSResourceCache` is not discussed in this RFC and the `alternate proposals` section says:
> If there is a desire to explicitly update the local cache (like apt), **we can introduce a Update-PSResourceCache cmdlet** with property on a PSRepository indicating whether it auto-updates or not. This would not be a breaking change.

So is use of `Update-PSResourceCache` here suggesting that we will introduce this cmdlet for sure? Are we going to have separate RFCs for the specifics of this cmdlet?

Reply:
Originally, the intent was to follow `apt` and make updating the cache explicit.  However, I've since agreed that making the cache update implicit is a better user experience.  This statement is just acknowledging that if it turns out that some users prefer explicit update, we can enable that by adding this cmdlet.